src/Nethermind/Nethermind.Synchronization/StateSync/StateSyncDispatcher.cs

```diff
@@ -32,11 +38,42 @@ public StateSyncDispatcher(ISyncFeed<StateSyncBatch> syncFeed, ISyncPeerPool syn
         {
         }
 
-        protected override async Task Dispatch(PeerInfo peerInfo, StateSyncBatch request, CancellationToken cancellationToken)
+        protected override async Task Dispatch(PeerInfo peerInfo, StateSyncBatch batch, CancellationToken cancellationToken)
         {
+            if (batch?.RequestedNodes == null || batch.RequestedNodes.Length == 0)
+            {
+                return;
+            }
+
             ISyncPeer peer = peerInfo.SyncPeer;
-            var getNodeDataTask = peer.GetNodeData(request.RequestedNodes.Select(n => n.Hash).ToArray(), cancellationToken);
-            await getNodeDataTask.ContinueWith(
+            Keccak[]? a = batch.RequestedNodes.Select(n => n.Hash).ToArray();
+            Task<byte[][]> task = null;
+
+            // Use GETNODEDATA if possible
+            if (peer.Node.EthDetails.Equals("eth66"))
+            {
+                task = peer.GetNodeData(a, cancellationToken);
+            }
+            // GETNODEDATA is not supported so we try with SNAP protocol
+            else if (peer.TryGetSatelliteProtocol("snap", out ISnapSyncPeer handler))
+            {
+                if (batch.NodeDataType == NodeDataType.Code)
+                {
+                    task = handler.GetByteCodes(a, cancellationToken);
+                }
+                else
+                {
+                    GetTrieNodesRequest request = GetGroupedRequest(batch);
+                    task = handler.GetTrieNodes(request, cancellationToken);
+                }
+            }
+
+            if (task is null)
+            {
+                return;
```

Comment:
NP: print warn.

Reply:
Also, probably should filter the peer in peer allocation to prevent this codepath. Otherwise the sync peer might just realloc this peer again immediately, causing a temporary cpu spike.